contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -4627,4 +4635,93 @@ static Node* optimize_explicit_cast(ParseState *pstate, Node *node)
 		}
 	}
 	return node;
-}
\ No newline at end of file
+}
+
+/* 
+ * Portals caches the entire or remaining result set on transaction 
+ * boundaries. Transaction commands will fail if this portal fetch
+ * fails. To handle this we silently fail and close SCROLL cursors
+ * although for NO_SCROLL cursors we make do with whatever is fetched.
+ * We also need to restore to the correct interrupt count value as we
+ * could already be inside a interrupt holdoff section. Also we do not
+ * rethrow the error here and end up in a dirty state, so some 
+ * transaction clean up should be done after this.
+ */
+static void
+pltsql_persist_holdable_cursor_executor(Portal portal, QueryDesc *queryDesc,
+										ScanDirection direction)
+{
+	uint32 saveInterruptHoldoffCount = InterruptHoldoffCount;
+	uint32 saveQueryCancelHoldoffCount = QueryCancelHoldoffCount;
+	int	save_pltsql_non_tsql_proc_entry_count = pltsql_non_tsql_proc_entry_count;
+	int	save_pltsql_sys_func_entry_count= pltsql_sys_func_entry_count;
+	
+	/* do this for pltsql open cursors only */
+	if (portal->name != NULL && pltsql_is_cursor_open(portal->name))
+	{
+		int          	prev_spi_stack_depth = SPI_get_depth();
+		MemoryContext 	old_context = CurrentMemoryContext;
+		
+		PG_TRY();
+		{
+			ExecutorRun(queryDesc, direction, 0L, false);
+		}
+		PG_CATCH();
+		{
+			MemoryContext ectx;
+			ErrorData   	*error_data;
+			CmdType		operation;
+			int     	current_spi_stack_depth = SPI_get_depth();
+			bool    	sendTuples;
+			int     	rc;
+			
+			/* implicitly close SCROLL type cursors on error */
+			if (portal->cursorOptions & CURSOR_OPT_SCROLL)
+			{
+				MarkPortalFailed(portal);
+				if (portal->portalPinned)
+					UnpinPortal(portal);
+				pltsql_update_cursor_state(portal->name, false);
+			}
+			else
+			{
+				ectx = MemoryContextSwitchTo(old_context);
+				error_data = CopyErrorData();
+				MemoryContextSwitchTo(ectx);
+				pltsql_update_cursor_error_message(portal->name, error_data->message, error_data->sqlerrcode);
+				FreeErrorData(error_data);
+			}
+
+			pop_top_error_stack();
+			
+			operation = queryDesc->operation;
+			
+			sendTuples = (operation == CMD_SELECT ||
+				queryDesc->plannedstmt->hasReturning);
+			
+			if (sendTuples)
+				queryDesc->dest->rShutdown(queryDesc->dest);
+
+			if (queryDesc->totaltime)
+				InstrStopNode(queryDesc->totaltime, queryDesc->estate->es_processed);
+
+			/* cleanup left over spi connections */
+			current_spi_stack_depth = SPI_get_depth();
+				
+			while (current_spi_stack_depth-- > prev_spi_stack_depth)
+				if ((rc = SPI_finish()) != SPI_OK_FINISH)
+					elog(ERROR, "SPI_finish failed: %s", SPI_result_code_string(rc));	
+
+			MemoryContextSwitchTo(old_context);
+
+			InterruptHoldoffCount = saveInterruptHoldoffCount;
```

Comment:
how do we make sure that we are restoring all states required? these might be few which were flagged in current use case but there could be more.

Reply:
These we restored since these are specific to babelfish i.e. we are setting them.
For other states we have to rely on the internal transaction command. So we have to make sure this function call is followed by some commit or abort transaction command (ones which actually do cleanup).
If we are not aborting or commiting afterwards then we do not need to hold the pin portals in the first place.

Reply:
For commit scenarios I think we might need to execute this inside a internal sub transaction to handle the clean up.
And error handling, abort transaction can handle the clean up.